### ISSUE TITLE
[Bug]: Transfer income does not add to cash

### ISSUE THREAD

**Issue by reedtanguerra** [OpenTTD/OpenTTD#10652]
### Version of OpenTTD

13.1, Windows 10 Home

### Expected result

Transfer income should increase cash and appear as income in the financial window.

### Actual result

The amount that should be received floats up as a text notification like usual, but cash does not increase and there is no income recorded in the financial window. The income does appear to be tabulating correctly in the vehicle group window. I normally play with NewGRFs but removed all of them to test if one of them was causing the issue. No, issue occurs regardless of mod status. Attached is a vanilla test game that makes no income at all.

[Transfer Income Vanilla Test.zip](https://github.com/OpenTTD/OpenTTD/files/11228818/Transfer.Income.Vanilla.Test.zip)

### Steps to reproduce

Build any transport route, and use transfer as one leg of the route.

**Comment by James103** [comment]
This is most likely intentional, as if transfer income was actually paid out to the company's available funds at the time of transfer, there arises the possibility of your company actually losing money when final delivery of cargo is made, not just the vehicle making the delivery.

**Comment by Eddi-z** [comment]
This is by design, transfer credits aren't actually paid until the cargo reaches its final destination.

**Comment by telk5093** [comment]
That's inteded behavior

**Comment by 2TallTyler** [comment]
As all have said, this is intentional and not a bug. 🙂

**Comment by reedtanguerra** [comment]
That's nonsensical. Services rendered should be paid as provided, but this is a man's game. :)

**Comment by TrueBrain** [comment]
Service rendered are paid .. when the leg finishes :) So as soon as you make your last leg, one that isn't a transfer, the whole leg gets paid what was due.

It is basically to prevent people only doing transfers, and not actually finishes a route. Just imagine, a route A -> B -> A, and you only do transfers between them. Just bringing people endless between two points, never releasing them to their final destination. They would see the same route, every day, unable to go home or enjoy the beach. The horror.

So basically, we wanted to prevent that kind of behaviour. So you get paid when you actually deliver. Not for transfers.

PS: this is not uncommon in real life too btw. Here for example I can buy a train-ticket from A to B to C, where A -> B are from one provider, and B -> C of another. I pay only one of the providers, and they distribute the fair share to the other.

So it is not that nonsensical .. I think :)

**Comment by reedtanguerra** [comment]
Thanks for the explanation. I see what you're trying to prevent (though why anyone would want to play in such a anti-fun way is beyond me). 

You're thinking of it as the consumer with your train ticket example. From the provider's perspective, the middle leg provider can bill the main carrier for services provided as soon as they've done it. It's called a performance obligation in accounting parlance, and once you've completed your contractual performance obligation, revenue recognition can take place. 

It also happens to irritate the shit out of me that I can have a ferry boat line making 20K a year according to the vehicle group window, but that income goes to road income because the two final destinations are reached by bus from the docks, while the maintenance of the ships still appears under the ships. That's a violation of the matching principle, which is an accounting rule that matches revenue with the associated expenses incurred to generate them (you guys are getting crash course in accounting from an rl accountant today)

**Comment by TrueBrain** [comment]
Haha, educate us please!

And yes, I agree, the way the transfer routes are billed is a bit weird, and unclear to the user. You capable of coding too, by any chance? :D

**Comment by reedtanguerra** [comment]
Sorry. I'm a whiz with spreadsheets and financial statements but programming is not my lane. I wish I had the time to learn coding. :)

**Comment by glx22** [comment]
Originally transfer didn't exist at all, and when making a multimodal route (with forced unload) only the last leg was making money, all other vehicles didn't make profit at all.
Transfer was added to have a fake profit for these vehicles.

**Comment by JGRennison** [comment]
> It may be worth looking into how @JGRennison (JGR's Patch Pack) handles profit sharing between multiple companies when shared infrastructure is enabled.

To save some digging, it is here: https://github.com/JGRennison/OpenTTD-patches/commit/9000d74f3c80edb9d325bb6980f9391b14c9ab93 (with some minor follow ups).

**Comment by reedtanguerra** [comment]
Please correct me if I'm wrong, but total revenue for the route is calculated by some formula of distance travelled, speed of delivery and cargo rate for the specific cargo, yes? That's why we have the cargo rate graph. Is it possible to do something like: boat travelled 10 miles at 10mph, and two buses did 1 mile each at 25mph for this route, split revenue money between boat and buses by distance times speed travelled or however the game calculates it? Revenue for all legs is only credited to cash when cargo arrives at final destination in keeping with your original purpose of deterring gameplay shenanigans?

**Comment by reedtanguerra** [comment]
On second thought, ignore the above. You'd get total game-backend processing insanity if the player uses a hub type system where cargo could be sitting at the central hub station for a while.

**Comment by PeterN** [comment]
Distance travelled is also direct from source to destination, ignoring any detours in between.

**Comment by Eddi-z** [comment]
the rough way it works is like this:
 * each "day"(*) a cargo spends inside a vehicle, an "age" counter is increased
 * at the final delivery, this "age" counter is matched against the distance from source to destination ("straight" line), to determine average speed, to calculate the payment
 * time spent at intermediate/hub stations is ignored

(*) not really a date, originally this was every 185 ticks (2.5 days), but can be modified by NewGRFs
